ESPS/general/src/dec_audio/decrecord.1
.\" Copyright (c) 1996 Entropic Research Laboratory, Inc.; All rights reserved
.\" @(#)decrecord.1	1.3 11/20/96 
.TH DECRECORD 1\-ESPS 11/20/96
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Research Lab., Inc.
.SH NAME
decrecord - mono or stereo record to disk
.PP
.SH SYNOPSIS
.B decrecord
[
.BI \-s " duration_to_record"
] [
.BI \-f " sample_rate"
] [
.BI \-c " number_of_channels"
] [
.BI \-S
] [
.BI \-P
] 
.I filename
.PP
.SH DESCRIPTION
.PP
.PP
\fIdecrecord\fP provides mono or stereo sampling at selected rates up
to 
48kHz when
recording to local disk.  Direct recording onto network disks is often
feasible as well.  Output files have ESPS FEA_SD headers.
\fIDecrecord\fP has special adaptations that permit display of the
recorded
file in \fIxwaves\fP (see \-S option below).
.PP
This program is for the DEC Alpha with multimedia extensions. To be
more precise, the DEC utility audiorecord must exist, because
\fIdecrecord\fP(1-ESPS) is simply a wrapper around it. The DEC
supplied audio capabilities are typically in /usr/bin/mme.
.PP
Use the DEC tool "audiocontrol" in the DEC multimedia extensions kit
to set the input and output controls (level, balance, port).  
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-s " duration" " [5]"
Specifies the maximum duration of the recording session in
seconds. Note that the record duration is specified as an integer, and
truncation of the value takes place if specified as a non-integer.
(That is, \fI\-s 1.9\fP gets interpreted as record for 1 second.)
.TP
.BI \-f " frequency" " [8000]"
Specifies the sampling frequency.  Only a limited set of sample rates
are available:
8000, 11025, 16000, 22050, 32000, 33075, 44100, and 48000.
Specifying a value other than one of these is an error.
.TP
.BI \-c " channel" " [1]"
By default \fIdecrecord\fP(1-ESPS) samples and stores single-channel data
from the left channel. Dual-channel (stereo) recording is selected by 
setting \fIchannel\fP
to 2.
When stereo recording
is selected channels 0 and 1 alternate in the file, with
channel 0 being first.  
.TP
.BI \-S
This option enable the \fIxwaves\fP (1-ESPS) "make" command via 
\fIsend_xwaves\fP (1-ESPS) after the
requested recording time has elapsed.
This permits immediate examination of the recorded passage using
\fIxwaves\fP (1-ESPS).
.TP
.BI \-P
This option enable shell prompting for the start of recording.
.PP
.SH ESPS PARAMETERS
.PP
The parameter file is not read.  
.PP
.SH ESPS COMMON
.PP
ESPS Common is not read or written.
.PP
.SH WARNINGS
.PP
decrecord does not work if another process has control of the audio
device, such as DECsound (decsound) or the VU meter of Audio Control.
You will get the following errors
.nf
    
    audiorecord: <error> Error opening Wave device
    audiorecord: <error> Unsupported wave format
.fi
.PP
The maximum rate over the network is unpredictable in general, but we
routinely achieve 16kHz stereo at Entropic Research Laboratory.  Of
course rate limitations due to network speed will be less severe for
single-channel recording.
.PP
.SH FILES
.PP
.SH BUGS
.PP
.PP
.SH SEE ALSO
.PP
\fIcopysd\fP(1\-\s-1ESPS\s+1), \fIdecplay\fP (1\-\s-1ESPS\s+1), \fIsfconvert\fP (1\-\s-1ESPS\s+1)
.PP
.SH AUTHOR
.PP
David Burton








ESPS/general/src/dec_audio/decrecord.sh
#!/bin/sh
#-------------------------------------------------------------------
#
#  This material contains proprietary software of Entropic Research Lab., Inc. 
#  Any reproduction, distribution, or publication without the the prior    
#  written permission of Entropic Research Lab., Inc. is strictly prohibited.
#  Any public distribution of copies of this work authorized in writing by
#  Entropic Research Lab., Inc. must bear the notice                    
#                                                               
#   "Copyright (c) 1996, 1997 Entropic Research Lab, Inc.; All rights reserved"
#                               
#
#  Sccs info: @(#)decrecord.sh	1.7 1/21/97 ERL
#  Written by: David Burton
#
# Brief Descrption: 
#
# This cover script uses the DEC Alpha supplied program
# audiorecord to record data and then converts the
# recorded data to an ESPS file by using SOX (the 
# SOund eXchange utility) and btosps(1-ESPS). 
#
# The audiorecord program cannot write to standard output, so
# we use a temporary file (and remove it when we are done).
# The audio hardware only seems to support sample rates that
# are multiples of 8000 or 11025, but not 24000. 
#
# The record program is not too precise about the number
# of samples it records (1 second at 11025 = 11032 samples),
# and the duration you specify is truncated to an integer.
# So, for example, if you specify a time of 1.9 seconds, you
# only record for 1 second.
#-------------------------------------------------------------------

#
# Turn of ESPS stuff
#

USE_ESPS_COMMON=off 
export USE_ESPS_COMMON
ESPS_VERBOSE=0
export ESPS_VERBOSE

# Set signal trap to exit on interrupt and clean up.
# We need this because the audiorecord program uses
# Cntrl-c to pause the recording, and we don't support this.
# 

Rpid=$$
trap 'kill -INT $Rpid;kill -INT $Rpid;sleep 1' 1 2 3 15

#
# Specify location of DEC audio utilities and SOX
#

DEC_AUDIO_BIN=/usr/bin/mme
DEC_AUDIO_RECORD=audiorecord
if test ! -x $DEC_AUDIO_BIN/$DEC_AUDIO_RECORD
then
	echo decrecord error:
	echo The DEC Alpha audio utilities are no longer in \"$DEC_AUDIO_BIN\",
	echo or the Alpha record program is no longer \"$DEC_AUDIO_RECORD\".
	echo The script decrecord needs to be fixed.
	exit 1
fi
SOX_BIN=$ESPS_BASE/bin


#
# Specify temp file directory
#

if test x$ESPS_TEMP_PATH != x
       	then
        tmpDir=$ESPS_TEMP_PATH
elif test x$TEMP != x
        then
        tmpDir=$TEMP
else
        tmpDir=/usr/tmp
fi


#
# Initialize variables
#

ShowInXwaves=0
Rate=8000
Duration=5
Channels=1
RecordPrompt=0


#
# Process command line arguments
#

while test x$1 != x
do
        case "$1" in
        -s) Duration=$2; shift;;
        -s*) Duration=`echo $1 | awk '{print substr($0, 3)}'` ;;

        -c) Channels=$2; shift;;
        -c*) Channels=`echo $1 | awk '{print substr($0, 3)}'` ;;

        -f) Rate=$2; shift;;
        -f*) Rate=`echo $1 | awk '{print substr($0, 3)}'` ;;

        -P) RecordPrompt=1;;

        -S) ShowInXwaves=1;;

        -?*) echo "decrecord: unknown option $1" ; exit 1 ;;
        *) break;;
        esac
        shift
done

# There should only be a file name left

if test $# -ne 1
        then
        echo \
	'Usage: decrecord [-P] [-S] [-s seconds] [-c channels] [-f rate] filename'
        exit 1
fi


#
# Set output file name
#

Outfile=$1


#
# Is output file writable?
# Create file, if it doesn't exist. If this fails, or
# the file exists and is not writable, warn and exit.
#

touch  $Outfile 2> /dev/null 
if test ! $? -eq 0
	then
	echo\
    "decrecord: Could not create output file ($Outfile) - check permissions.\n"
	exit 1
fi

if test ! -w $Outfile
	then
	echo\
    "decrecord: Could not create output file ($Outfile) - check permissions.\n"
	exit 1
fi


#
# Is number of channels valid?
#

case $Channels in
	[1-2])
		# things OK
		;;
	*) 	
		echo \
		  "decrecord: Invalid number of channels specified: $Channels."
		echo "           Only 1 or 2 channels are supported.\n";
		exit 1;;

esac


#
# Is sample rate valid?
#

case $Rate in
	
	# Following rates OK
	8000) 		;;
	11025)		;;
	16000)		;;
	22050)		;;
	32000)		;;
	33075)		;;
	44100)		;;
	48000)		;;

	24000)
		echo "decrecord: Invalid sample rate specified: $Rate."
		echo "     Try a multiple of 8000 or 11025 (but not 24000).\n";
		exit 1;;
		

	*)
		echo "decrecord: Invalid sample rate specified: $Rate."
		echo "     Try a multiple of 8000 or 11025 (but not 24000).\n";
		exit 1;;		
esac


#
# Is duration valid?
#

if test $Duration -lt 1
	then
	echo "decrecord: Invalid record duration specified: $Duration"
	echo "           Duration must be an integer and >= 1"
	exit 1
fi


#
# Record data and convert to an an ESPS file.
# We put audiorecord in background and save the process ID, so we can
# kill it on clean up, if the user interrupts.
#

$DEC_AUDIO_BIN/$DEC_AUDIO_RECORD -encoding pcm -bitspersample 16 -rate $Rate\
	-channels $Channels -time $Duration \
	-filename $tmpDir/drecord$$ >/dev/null &

#
# save process ID 
#

Rpid=$!

#
# Prompt after record begins to allow hard a chance to initialize
#

if test $RecordPrompt -eq 1
	then
	echo "BEGIN RECORDING NOW..."
fi

#
#  wait for record to finish
#

wait $Rpid

if test $RecordPrompt -eq 1
	then
	echo "RECORDING DONE"
fi


# Convert to ESPS file

$SOX_BIN/sox -t wav $tmpDir/drecord$$ -t raw - | $ESPS_BASE/bin/btosps \
	-f $Rate -n $Channels -t short \
	-c"Converted via $DEC_AUDIO_RECORD and SOX."\
	 - $Outfile


#
# clean up
#

rm -f $tmpDir/drecord$$ &


# If display in xwaves is specified, give it a try.
# Note that we do not do any checking to see if xwaves
# already exists. We use the send_xwaves error output
# to inform the user about the unavailability of xwaves.

if test $ShowInXwaves -eq 1
	then
	# if full path given  (check for leading "/")
	# then
	#	use it
	# else
	# add path so that xwaves can find file

	if echo $Outfile | grep -q \^/
	then 
		send_xwaves -D 0 make file $Outfile
	else
		send_xwaves -D 0 make file `pwd`/$Outfile
	fi
fi



